Fixes evil-mode dragging and adds evil-mode recipes

gabrielsimoes opened the pull request:
This uses @luxbock fix from https://github.com/rejeep/drag-stuff.el/pull/10 and adds recipes for testing evil-mode.

gabrielsimoes commented:
Well, all checks were working on my machine (emacs 24.5.1), I don't think it's evil-mode's fault.

gabrielsimoes commented:
Fixed, checks are working now

gabrielsimoes commented:
Any update? Please don't let this project die.

rejeep commented on features/conflicting-modes.feature:
```diff
@@ -96,3 +96,51 @@ Feature: Drag Stuff
     Then longlines-mode should not be active
+
+  Scenario Outline: Evil region left/right
+	Given I insert:
```
Invalid indentation

rejeep commented on drag-stuff.el:
```diff
@@ -183,9 +187,12 @@
          (mark-line (line-number-at-pos (mark)))
-         (point-line (line-number-at-pos (point)))
+         (point-line (line-number-at-pos
+                      (if evilp (evil-visual-goto-end) (point))))
```
Can you please explain this change as there is a side effect in evil, but not in default.

gabrielsimoes replied:
I still have to read the code, actually I just got the changes from @luxbock pull request and added test recipes. However, I noticed evil-mode also selects the lower line each time you drag up or down. So it ends up draggin more than it should every time. This change fixes it, but makes it unable to drag down to the last line of the file. I will try to provide a better solution.

rejeep commented:
@gabrielsimoes Really sorry about this. I don't know how, but I didn't notice this pull request when it was first sent.

I made a few comments in the code. Can you please address them and when done squash your commits into a single one.

gabrielsimoes commented:
No problem, sorry if it seemed rude or anything. I'll try to address your comment as soon as I can.

gabrielsimoes commented on drag-stuff.el:
```diff
@@ -167,7 +171,7 @@
 (defun drag-stuff-lines-up (arg)
   "Move all lines in the selected region ARG lines up."
-  (if (> (line-number-at-pos (min (point) (mark))) (abs arg))
+  (if (> (line-number-at-pos (region-beginning)) (abs arg))
```
I thought this was a better solution, if you prefer the other way, I may roll back this change.

gabrielsimoes commented on drag-stuff.el:
```diff
@@ -175,22 +179,34 @@
 (defun drag-stuff-lines-down (arg)
   "Move all lines in the selected region ARG lines up."
-  (if (<= (+ (line-number-at-pos (max (point) (mark))) arg) (count-lines (point-min) (point-max)))
+  (let ((lastline (if (and (drag-stuff--evil-p)
```
Evil visual selection mode has a strange behaviour: If you press <End> it actually doesn't move to the last char of the line, but the first char of the next line. So, this checks if this is happening, and goes back one char before testing.

gabrielsimoes commented:
I thought I had fixed it, but some new bugs appeared.

gabrielsimoes commented:
Alright, I wrote a better solution and squashed the commits. You may want to change something.

rejeep commented on drag-stuff.el:
```diff
@@ -167,30 +171,44 @@
 (defun drag-stuff-lines-down (arg)
   "Move all lines in the selected region ARG lines up."
-  (if (<= (+ (line-number-at-pos (max (point) (mark))) arg) (count-lines (point-min) (point-max)))
+  (let ((lastline (if (drag-stuff--evil-p)
```
The name `lastline` is a bit misleading as it's not a line. It's the point at the end of the region. So let's call it something like `region-end` or similar.

rejeep commented on drag-stuff.el:
```diff
@@ -167,30 +171,44 @@
+                                            (current-column)))
+         (point-col (if (and evilp (> (point) (mark)))
+                      (save-mark-and-excursion (evil-visual-goto-end) (current-column))
+                      (current-column)))
```
I think all these should be in private (functions starting with `drag-stuff--`) functions. For example:

* `drag-stuff--mark-line`
* `drag-stuff--point-line`
* `drag-stuff--mark-col`
* `drag-stuff--point-col`

rejeep commented on drag-stuff.el:
```diff
@@ -225,12 +247,17 @@
-    (if mark-active
-        (exchange-point-and-mark))
-    (setq end (line-end-position))
+    (if evilp
```
Use `cond`

rejeep commented on drag-stuff.el:
```diff
@@ -225,12 +247,17 @@
-    (setq end (line-end-position))
+    (if evilp
+      (progn (setq beg (save-mark-and-excursion (goto-char (region-beginning)) (line-beginning-position)))
+             (setq end (save-mark-and-excursion (evil-visual-goto-end) (line-end-position))))
```
As we want both beg and end, I think it would make sense to use `evil-visual-range` here instead.

gabrielsimoes replied:
Didn't know about it, but actually `evil-visual-range` is affected by evil's behaviour on selecting one more char than it should when selecting the end of a line, but `evil-visual-goto-end` isn't.

gabrielsimoes commented:
Great points, I also aligned let* declarations at `drag-stuff-lines-vertically`.

rejeep commented on drag-stuff.el:
```diff
@@ -167,30 +195,35 @@
+         (beg        (car bounds))
+         (end        (car (cdr bounds)))
+         (deactivate-mark nil))
+
```
Please don't add whitespace like this. As you can see, it's impossible to see what changed in this diff.

rejeep commented on drag-stuff.el:
```diff
@@ -89,13 +89,41 @@
                (list drag-stuff-modifier))))
     (vector (append mod (list key)))))
 
+(defun drag-stuff--mark-line ()
```
I know it was my idea, but can you change the name of the functions to:

* `drag-stuff--line-at-mark`
* `drag-stuff--line-at-point`
* `drag-stuff--col-at-mark`
* `drag-stuff--col-at-point`

I didn't realise att first that the name `drag-stuff--mark-line` sound like it marks a line. :smile:

gabrielsimoes commented:
Ok, I renamed the functions, and rewrote them with better evil's variables for `point` and `mark`. With this rewrite, and also an addition at the end of `drag-stuff-lines-vertically`, the cursor location after dragging lines with evil's line selection is fixed (just a minor bug). I also removed the spaces. I won't ammend the commits until merging this.

gabrielsimoes commented:
I plan on adding tests for the cursor position later, but I think it is better to keep it for another pull request.

rejeep commented:
> I plan on adding tests for the cursor position later, but I think it is better to keep it for another pull request.

Then the feature should be in a separate PR as well. So either add the test in this PR or fix that bug (with test) in another PR.

gabrielsimoes commented:
Sure, I prefer adding the tests here. But I will only add tests for evil mode's behaviour, and keep cursor tests for emacs in general out of this PR.

gabrielsimoes commented:
I think I got a little confused on my last two comments here. :sweat_smile: Anyway, I wrote tests for all the things evil-mode might behave differently and are fixed in this PR, and placed everything at `evil-mode.feature`.

rejeep commented:
I will try out this branch locally for a little while and if everything works, I'll merge it.

rejeep commented:
Merged, thanks for your work (and patience) on this! 👍

gabrielsimoes commented:
Great! You're welcome

gabrielsimoes commented:
This closes #10 and #5
